lock-api/src/main/java/com/palantir/lock/client/TransactionStarter.java

```diff
@@ -138,16 +131,21 @@ public void close() {
     }
 
     @VisibleForTesting
-    static Consumer<List<BatchElement<Void, StartIdentifiedAtlasDbTransactionResponse>>> consumer(
+    static Consumer<List<BatchElement<Integer, List<StartIdentifiedAtlasDbTransactionResponse>>>> consumer(
             LockLeaseService lockLeaseService, LockWatchEventCache lockWatchEventCache) {
         return batch -> {
-            int numTransactions = batch.size();
+            int numTransactions = batch.stream().map(BatchElement::argument).reduce(0, Integer::sum);
 
             List<StartIdentifiedAtlasDbTransactionResponse> startTransactionResponses =
                     getStartTransactionResponses(lockLeaseService, lockWatchEventCache, numTransactions);
 
-            for (int i = 0; i < numTransactions; i++) {
-                batch.get(i).result().set(startTransactionResponses.get(i));
+            int start = 0;
+            int end;
+            for (BatchElement<Integer, List<StartIdentifiedAtlasDbTransactionResponse>> batchElement
+                    : batch) {
+                end = start + batchElement.argument();
+                batchElement.result().set(startTransactionResponses.subList(start, end));
+                start = end;
```

Comment:
This code is a bit messy, for sure - but more a proof of concept of how it would work.

Should definitely go in its own PR.